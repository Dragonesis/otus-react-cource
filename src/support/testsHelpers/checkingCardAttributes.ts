import { screen } from '@testing-library/react'
import { Unit } from '@/services/models'

export const checkingCardAttributes = (unit: Unit) => {
  const card = screen.queryAllByLabelText('Карта ' + unit.name)[0]
  expect(card).toBeInTheDocument()
  expect(screen.getByText(unit.name)).toBeInTheDocument()
  expect(screen.getByText('Сила: ' + unit.power)).toBeInTheDocument()
  expect(screen.getByText('Защита: ' + unit.defense)).toBeInTheDocument()
  expect(screen.getByText('Класс: ' + unit.className)).toBeInTheDocument()
}
