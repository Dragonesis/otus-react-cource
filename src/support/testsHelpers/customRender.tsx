import React from 'react'
import { render } from '@testing-library/react'
import { StoreContext } from '@/services/adapters/store'
import { DeckOfCard, Unit } from '@/services/models'

export const customRender = (
  children: React.ReactNode,
  {
    providerProps,
    ...renderOptions
  }: {
    providerProps: {
      cardDeck: DeckOfCard | undefined
      cardsInHand: Unit[]
      setDeckOfCard: (arg: DeckOfCard) => DeckOfCard
      setCardsInHand: (arg: Unit[]) => Unit[]
    }
  },
) => {
  return render(<StoreContext.Provider value={providerProps}>{children}</StoreContext.Provider>, renderOptions)
}
