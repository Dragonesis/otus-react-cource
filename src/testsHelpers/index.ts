export { customRender } from './customRender'
export { checkingCardAttributes } from './checkingCardAttributes'
