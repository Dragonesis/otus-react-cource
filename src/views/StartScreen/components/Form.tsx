import React, { FC, useCallback, useState, ChangeEvent } from 'react'
import { Field, FieldWrap, Button } from '@/ui'
import { User } from '@/services/models'
import styled from '@emotion/styled'
import { useStore } from '@/services/adapters/store'
import { deckOfCardList } from '@/services/mocks'
import { getDeckOfCard } from '@/application'

export interface FormProps {
  onMouseEnter: () => void
  onMouseLeave: () => void
  setUser: (arg: User) => void
}

export type FormErrors<Type> = {
  [Property in keyof Type]: boolean
}

export const Form: FC<FormProps> = ({ onMouseEnter, onMouseLeave, setUser }) => {
  const { setDeckOfCard, setCardsInHand } = useStore()
  const [values, setValues] = useState<User>({
    name: '',
    email: '',
    birthday: '',
  })
  const [errors, setErrors] = useState<FormErrors<User>>({
    name: false,
    email: false,
    birthday: false,
  })
  const [isBirthdayCalendar, setIsBirthdayCalendar] = useState<boolean>(false)

  const handlingOnChange = useCallback((e: ChangeEvent<HTMLInputElement>) => {
    setValues((preState) => {
      preState[e.target.name] = e.target.value
      return preState
    })
  }, [])

  const handlingOnSubmit = useCallback((e: ChangeEvent<HTMLFormElement>) => {
    e.preventDefault()
    const errors = [
      'name',
      'email',
      'birthday',
    ].reduce((acc: { [key: string]: boolean }, val) => {
      const el = e.target.elements.namedItem(val) as HTMLInputElement
      acc[val] = !el.validity.valid
      return acc
    }, {})

    setErrors(errors)
    if (!Object.values(errors).find(item => item) && values.name) {
      setUser(values)
      setDeckOfCard(getDeckOfCard(deckOfCardList))
      setCardsInHand([])
    }
  }, [])

  return (
    <Core onSubmit={handlingOnSubmit} noValidate>
      <FieldWrap isOffset>
        <Field
          isError={errors.name}
          required name='name' onChange={handlingOnChange} placeholder='Имя'/>
      </FieldWrap>
      <FieldWrap isOffset>
        <Field
          isError={errors.email}
          required name='email' type='email' onChange={handlingOnChange} placeholder='Email' />
      </FieldWrap>
      <FieldWrap>
        <Field
          isError={errors.birthday}
          required
          name='birthday'
          type={isBirthdayCalendar ? 'date' : 'text'}
          onFocus={() => setIsBirthdayCalendar(true)}
          onBlur={(e) => !e.target.value && setIsBirthdayCalendar(false)}
          onChange={handlingOnChange}
          placeholder='День рождения'
        />
      </FieldWrap>

      <Action
        size='l'
        fixSize='m'
        type='submit'
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        data-testid='action-in-games'
      >
        В игру
      </Action>
    </Core>
  )
}


const Core = styled.form`
  display: block;
  margin-top: 20px;
  max-width: 360px;
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
`
const Action = styled(Button)`
  margin-top: 20px;
`