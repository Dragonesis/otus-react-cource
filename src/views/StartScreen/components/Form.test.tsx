import React from 'react'
import { render, fireEvent, screen } from '@testing-library/react'
import { customRender } from '@/support/testsHelpers'

import { Form } from './Form'

describe('### Form', () => {
  const setDeckOfCard = jest.fn()
  const setCardsInHand = jest.fn()

  const providerProps = {
    cardDeck: undefined,
    cardsInHand: [],
    setDeckOfCard,
    setCardsInHand,
  }

  const setUser = jest.fn()
  const onMouseEnter = jest.fn()
  const onMouseLeave = jest.fn()

  test('Form renders correctly', () => {
    const { asFragment } = render(<Form setUser={setUser} onMouseEnter={onMouseEnter} onMouseLeave={onMouseLeave} />)
    expect(asFragment()).toMatchSnapshot()
  })
  test('Form submit', () => {
    const setUser = jest.fn()
    const onMouseEnter = jest.fn()
    const onMouseLeave = jest.fn()
    const { container } = customRender(<Form setUser={setUser} onMouseEnter={onMouseEnter} onMouseLeave={onMouseLeave} />, { providerProps })

    const name = screen.getByPlaceholderText('Имя')
    const email = screen.getByPlaceholderText('Email')
    const birthday = screen.getByPlaceholderText('День рождения')
    fireEvent.change(name, { target: { value: 'test' } })
    fireEvent.change(email, { target: { value: 'test@ya' } })
    fireEvent.change(birthday, { target: { value: '10.01.1999' } })

    const action = container.querySelector('[data-testid="action-in-games"]')
    expect(action).toBeInTheDocument()
    if (action) {
      fireEvent.click(action)
      expect(setUser).toHaveBeenCalled()
    }
  })
  test('Validation', () => {
    const { container } = customRender(<Form setUser={setUser} onMouseEnter={onMouseEnter} onMouseLeave={onMouseLeave} />, { providerProps })

    const action = container.querySelector('[data-testid="action-in-games"]')
    expect(action).toBeInTheDocument()
    if (action) {
      fireEvent.click(action)
      expect(setUser).not.toHaveBeenCalled()
    }
  })
})
