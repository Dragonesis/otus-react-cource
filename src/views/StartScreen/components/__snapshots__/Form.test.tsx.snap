// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`### Form Form renders correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: block;
  margin-top: 20px;
  max-width: 360px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-1 {
  display: block;
  width: 100%;
  margin-bottom: 16px;
}

.emotion-2 {
  -webkit-transition: border-color 0.2s ease-in-out;
  transition: border-color 0.2s ease-in-out;
  width: 100%;
  height: 36px;
  border-radius: 4px;
  border: 1px solid #3a3a3a;
  padding-left: 10px;
  padding-right: 10px;
  background-color: #fff;
}

.emotion-2:hover,
.emotion-2:focus {
  border-color: #9d9d9d;
}

.emotion-5 {
  display: block;
  width: 100%;
}

.emotion-7 {
  border-radius: 6px;
  width: 100%;
  max-width: 210px;
  height: 60px;
  padding-left: 16px;
  padding-right: 16px;
  font-size: 20px;
  background-color: #3a3a3a;
  color: #fff;
  margin-top: 20px;
}

.emotion-7:hover,
.emotion-7:focus {
  border-color: #303030;
}

.emotion-7.size-s {
  height: 36px;
  padding-left: 10px;
  padding-right: 10px;
  font-size: 16px;
}

.emotion-7.size-m {
  height: 48px;
  padding-left: 10px;
  padding-right: 10px;
  font-size: 18px;
}

.emotion-7.size-l {
  height: 60px;
  padding-left: 16px;
  padding-right: 16px;
  font-size: 20px;
}

<form
    class="emotion-0"
    novalidate=""
  >
    <label
      class="emotion-1"
    >
      <input
        class="emotion-2"
        name="name"
        placeholder="Имя"
        required=""
      />
    </label>
    <label
      class="emotion-1"
    >
      <input
        class="emotion-2"
        name="email"
        placeholder="Email"
        required=""
        type="email"
      />
    </label>
    <label
      class="emotion-5"
    >
      <input
        class="emotion-2"
        name="birthday"
        placeholder="День рождения"
        required=""
        type="text"
      />
    </label>
    <button
      class="emotion-7"
      data-testid="action-in-games"
      type="submit"
    >
      В игру
    </button>
  </form>
</DocumentFragment>
`;
